src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java

```diff
@@ -1263,6 +1284,209 @@ public void visitMethodDef(JCMethodDecl tree) {
+                if (rs.isEarlyReference(
+                        true,
+                        localEnv,
+                        tree.meth instanceof JCFieldAccess fa ? fa.selected : null,
+                        msym))
+                    reportPrologueError(tree.meth, msym);
+            }
+        }
+
+        @Override
+        public void visitIdent(JCIdent tree) {
+            // skip if this identifier is part of a select, context is important here
+            if (!analyzingSelect) {
+                analyzeSymbol(tree);
+            }
+        }
+
+        boolean analyzingSelect = false;
+
+        @Override
+        public void visitSelect(JCFieldAccess tree) {
+            // skip if part of a larger select, context is important here
+            if (!analyzingSelect) {
+                boolean previousAnalyzingSelect = analyzingSelect;
+                try {
+                    analyzingSelect = true;
+                    super.visitSelect(tree);
+                    analyzeSymbol(tree);
+                } finally {
+                    analyzingSelect = previousAnalyzingSelect;
+                }
+            }
+        }
+
+        @Override
+        public void visitNewClass(JCNewClass tree) {
+            super.visitNewClass(tree);
+            checkNewClassAndMethRefs(tree, tree.type);
+        }
+
+        @Override
+        public void visitReference(JCMemberReference tree) {
+            super.visitReference(tree);
+            if (tree.getMode() == JCMemberReference.ReferenceMode.NEW) {
+                checkNewClassAndMethRefs(tree, tree.expr.type);
+            }
+        }
+
+        void checkNewClassAndMethRefs(JCTree tree, Type t) {
+            if (t.tsym.isEnclosedBy(localEnv.enclClass.sym) &&
+                    !t.tsym.isStatic() &&
+                    !t.tsym.isDirectlyOrIndirectlyLocal()) {
+                reportPrologueError(tree, t.getEnclosingType().tsym);
+            }
+        }
+
+        /* if a symbol is in the LHS of an assignment expression we won't consider it as a candidate
+         * for a proxy local variable later on
+         */
+        boolean isInLHS = false;
+
+        @Override
+        public void visitAssign(JCAssign tree) {
+            boolean previousIsInLHS = isInLHS;
+            try {
+                isInLHS = true;
+                scan(tree.lhs);
+            } finally {
+                isInLHS = previousIsInLHS;
+            }
+            scan(tree.rhs);
+        }
+
+        @Override
+        public void visitMethodDef(JCMethodDecl tree) {
+            // ignore any declarative part, mainly to avoid scanning receiver parameters
+            scan(tree.body);
+        }
+
+        void analyzeSymbol(JCTree tree) {
+            tree = TreeInfo.skipParens(tree);
+            Symbol sym = TreeInfo.symbolFor(tree);
+            if (sym != null) {
+                if (!sym.isStatic() && !isMethodArgument(tree)) {
```

Comment:
if you have a `sym`, in order to understand if something is a method parameter (not argument?) don't you need to check if `sym.owner == MTH` ?

Reply:
this is for cases when we have an argument that is for example of the same type as the current class so like:

class Test {
    String s;
    
    Test(Test t) {
        // the owner of s is Test not MTH so we need to check what is the qualifier for s which at the end is the argument
        // `t` so we ignore it
        String s1 = t.s;
        super();
    }
}

Reply:
I tried to replace this with:

`if (!sym.isStatic() && sym.kind == VAR && sym.owner.kind == TYP) { ... }`

And no tests seem to fail.